### PR TITLE
[WIP] Parametric Logarithmic Derivative: using Risch structure Theorem

### PR DESCRIPTION
**Pull request by gxyd**

### CODE REVIEW

File: sympy/integrals/prde.py [hunk 1]
```diff
@@ -1038,6 +1042,28 @@ def is_deriv_k(fa, fd, DE):
...
+    from sympy import sfield, AlgebraicField
+
+    K = AlgebraicField(QQ, *G.domain.gens)
+    trans_nums = list(set(G.gens) - set([G.gen]))
+    expr_args = G.as_expr().args
+    # T = [t1, t2, ..., tn]
+    T = set([num**(i.as_coeff_exponent(num)[1])
+        for num in trans_nums for i in expr_args])
+    cons = T.union(set(K.gens))
+    return sfield(list(cons), extension=True)
```
Comment (gxyd) [inline sympy/integrals/prde.py]: It raises error when the flag `extension=True` is there.

File: sympy/integrals/prde.py [hunk 2]
```diff
@@ -1038,6 +1049,47 @@ def is_deriv_k(fa, fd, DE):
...
+    """
+    from sympy import sfield, AlgebraicField, QQ
+    # algebracic constants
+    algebraic_consts = set()
+    # transcendental constants that are algebraically
+    # independent
+    alg_ind_trans_consts = set()
+
+    for poly in G:
+        trans_consts = list(set(poly.gens) - set([poly.gen]))
```
Comment (jksuom) [inline sympy/integrals/prde.py]: What should be assumed of `poly`. How many generators and which kind of `domain` should it have?

Comment (gxyd) [inline sympy/integrals/prde.py]: Atleast domain shouldn't be `EX`. It could be an instance of `AlgebraicField` or `PythonRationalField`.

As for the generators, I think it could be `pi` , `E` i.e. the transcendental numbers should need to be there (but the algebraic number not, as `extension=True` should be there with each poly).

Comment (jksuom) [inline sympy/integrals/prde.py]: I wonder why all generators except for the principal one (`poly.gen`) are taken. They could probably all be considered transcendental for computing the rational relations. (It probably does not matter whether they are constants or not in the differential extension. This question is purely algebraic.)

File: sympy/integrals/prde.py [hunk 3]
```diff
@@ -918,14 +918,25 @@ def parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1=None):
...
-        # It isn't too worrisome, because the heuristic handles most difficult
-        # cases.
+    """
+    Page, 251 section 7.3
+    """
+
+    C, K, trans_consts = base_Q(DE.D)
+    # trans_consts be t1, ..., tn
+    E_part = [frac_in(DE.D[i].quo(Poly(DE.T[i], DE.T[i])),
+        trans_consts) for i in DE.indices('exp')]
```
Comment (jksuom) [inline sympy/integrals/prde.py]: What is the result of `frac_in` when `trans_consts` is a list?

File: sympy/integrals/prde.py (hunk 2 above)
Comment (gxyd) [inline sympy/integrals/prde.py]: > I wonder why all generators except for the principal one (poly.gen) are taken.

Here "all generators" -> "transcendental generators" ?
As for not including `poly.gen`, I think you are right, I should include it. But I think I haven't found a good enough test case for this function. I'll try to do that first and see if it works okay.

File: sympy/integrals/prde.py (hunk 3 above)
Comment (gxyd) [inline sympy/integrals/prde.py]: An example would be
```python
In [4]: frac_in(x**2/(x + 1), [x, t, sqrt(2)])
Out[4]: (Poly(x**2, x, t, sqrt(2), domain='ZZ'), Poly(x + 1, x, t, sqrt(2), domain='ZZ'))
```
So it uses all the elements of list as generator for `num` and `den`.

**Comment by gxyd** [comment]
I have updated the PR. I have also added one doctest example, showing how things work for now with the function `parametric_log_deriv`. (I wouldn't update the PR if it bothers you in the PR reviewing, while you are doing it).

File: sympy/integrals/prde.py [hunk 4]
```diff
@@ -1038,6 +1062,47 @@ def is_deriv_k(fa, fd, DE):
...
+    """
+    from sympy import sfield, AlgebraicField, QQ
+    # algebracic constants
+    algebraic_consts = set()
+    # transcendental constants that are algebraically
+    # independent
+    alg_ind_trans_consts = set()
+
+    for poly in G:
+        trans_consts = list(set(poly.gens))
```
Comment (jksuom) [inline sympy/integrals/prde.py]: Could this be `poly.gens`?

File: sympy/integrals/prde.py [hunk 5]
```diff
@@ -1038,6 +1062,47 @@ def is_deriv_k(fa, fd, DE):
...
+        expr_args = poly.as_expr().args
+        # T = [t1, t2, ..., tn]
+        # obtaining algebraically independent
+        # for ex. if π and π^2 both are in poly.as_expr() then
+        # both should be added
+        T = set([num**(i.as_coeff_exponent(num)[1])
+            for num in trans_consts for i in expr_args])
```
Comment (jksuom) [inline sympy/integrals/prde.py]: These seem to be the monomials as expressions. Would it be possible to use `monoms`?

File: sympy/integrals/prde.py (hunk 4 above)
Comment (gxyd) [inline sympy/integrals/prde.py]: Yes, I think that would be okay.

File: sympy/integrals/prde.py (hunk 5 above)
Comment (jksuom) [inline sympy/integrals/prde.py]: I did mean the `.monoms()` method of polynomials that does not require transforming to expressions.

File: sympy/integrals/prde.py [hunk 6]
```diff
@@ -214,24 +215,24 @@ def poly_linear_constraints(p, d):
...
+    has no constant solution.
 
     This algorithm is used both in solving parametric problems and in
-    determining if an element a of K is a derivative of an element of K or the
-    logarithmic derivative of a K-radical using the structure theorem approach.
+    determining if an element of K is a derivative of an element of K
+    or a logarithmic derivative of a K-radical using the structure
+    theorem approach.
 
-    Because Poly does not play well with Matrix yet, this algorithm assumes that
```
Comment (asmeurer) [inline sympy/integrals/prde.py]: Why did you remove this? Is it not true any more?

File: sympy/integrals/prde.py [hunk 7]
```diff
@@ -1038,6 +1063,45 @@ def is_deriv_k(fa, fd, DE):
             return (ans, result, const)
 
 
+def base_Q(G):
+    """
+    >>> G = [Poly(x**2*sqrt(2) + x, x, extension=True),
```
Comment (asmeurer) [inline sympy/integrals/prde.py]: Please add function description to the docstring.

File: sympy/integrals/prde.py [hunk 8]
```diff
@@ -1038,6 +1063,45 @@ def is_deriv_k(fa, fd, DE):
             return (ans, result, const)
 
 
+def base_Q(G):
```
Comment (asmeurer) [inline sympy/integrals/prde.py]: Should be `basis_Q`, not `base_Q` (if this does what it looks like it does).

Comment (asmeurer) [inline sympy/integrals/prde.py]: Can you just use the "parallel" constructor in the polys to do this work for you?

File: sympy/integrals/prde.py [hunk 9]
```diff
@@ -1038,6 +1063,45 @@ def is_deriv_k(fa, fd, DE):
...
+    # algebracic constants
+    algebraic_consts = set()
+    # transcendental constants that are algebraically
+    # independent
+    alg_ind_trans_consts = set()
+
+    for poly in G:
+        monoms = poly.as_expr().args
+        # T = [t1, t2, ..., tn]
+        # obtaining algebraically independent
```
Comment (asmeurer) [inline sympy/integrals/prde.py]: What does "obtaining" mean here? 

And I don't understand this comment because π and π^2 are not algebraically independent.

File: sympy/integrals/prde.py (hunk 6 above)
Comment (gxyd) [inline sympy/integrals/prde.py]: We have `PolyMatrix` now, so I think we can (I haven't done it) modify the code of `constant_system` to receive input as tuple of Poly's `(poly_num, poly_den)`.

File: sympy/integrals/prde.py (hunk 9 above)
Comment (gxyd) [inline sympy/integrals/prde.py]: That is wrong what I did here, basically it was earlier calculating 'linearly' independent transcendental constants. I have improved the function to calculate 'algebraically' independent constants.

File: sympy/integrals/prde.py [hunk 10]
```diff
@@ -918,15 +919,46 @@ def parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1=None):
...
+
+    nums = [denom.exquo(f.denom)*f.numer for f in gens]
+    monoms = set()
+    for num in nums:
+        monoms.update(num.monoms())
+
+    # each monomial corresponds to a row in the matrix
+    # of entries in 'K' giving the relations.
+    # The elements of F correspond to the columns
+    t = [[num._get_coeff(m) for num in nums] for m in monoms]
```
Comment (gxyd) [inline sympy/integrals/prde.py]: I'll make this a `PolyMatrix` (imported as `Matrix`).

Comment (jksuom) [inline sympy/integrals/prde.py]: The matrix will be needed to find the nullspace after the entries are made rational.

File: sympy/integrals/prde.py [hunk 11]
```diff
@@ -918,15 +919,61 @@ def parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1=None):
...
+    if K.is_Algebraic:
+        # 'Kmat' is not the same as 'Qmat'
+        n = K.mod.degree()
+        zero = QQ.zero
+
+        def pad(rep):
+            return [zero]*(n - len(rep)) + rep
+
+        # replace each row of Kmat with n rows of Qmat
+        Qmat = Matrix([zip(*[pad(a.rep) for a in row]) for row in Kmat])
```
Comment (jksuom) [inline sympy/integrals/prde.py]: What is the type of `Qmat[0. 0]`?

Comment (gxyd) [inline sympy/integrals/prde.py]: Type of `Qmat[0, 0]` is `list` (it shouldn't be).

Comment (gxyd) [inline sympy/integrals/prde.py]: Better said to be a `iterator` (which iterates over a list).

Comment (jksuom) [inline sympy/integrals/prde.py]: That is the reason why I suggested a loop `for row in Kmat` where each time n new rows are generated the list is extended, not appended.

Comment (gxyd) [inline sympy/integrals/prde.py]: Yes, I figure that there must be something with that. I am guessing they are separately to be taken as `Matrix` (but please keep the details in hide for a moment).

File: sympy/integrals/prde.py [hunk 12]
```diff
@@ -944,7 +945,8 @@ def parametric_log_deriv(fa, fd, wa, wd, DE):
             F.append(DE.D[i].as_expr())
 
     # find a field containing all the fractions
-    FF, gens = sfield(F, extension=True)
+    # 'field=True' covers the non-algebraic case
+    FF, gens = sfield(F, field=True)
```
Comment (jksuom) [inline sympy/integrals/prde.py]: `extension=True` should also be there.

File: sympy/integrals/prde.py [hunk 13]
```diff
@@ -917,16 +918,96 @@ def parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1=None):
...
+    Examples
+    ========
+
+    """
+    # find a field containing all the fractions
+    # 'field=True' covers the non-algebraic case
+    FF, gens = sfield(F, field=True, extension=True)
+    K = FF.domain
+
+    denom = reduce(lambda i, j: i.lcm(j), [gen.denom for gen in gens])
```
Comment (jksuom) [inline sympy/integrals/prde.py]: `reduce` is no longer a built-in function in Python 3. It has to be imported from functools. (Its use is probably being discouraged.)

File: sympy/integrals/prde.py [hunk 14]
```diff
@@ -979,6 +992,9 @@ def pad(rep):
         # 'Kmat' is the same of 'Qmat'
         Qmat = Matrix(Kmat)
 
+    V = Qmat.rowspace()
```
Comment (jksuom) [inline sympy/integrals/prde.py]: The set of linear relations is given by the nullspace of Qmat.

Comment (gxyd) [inline sympy/integrals/prde.py]: Yes, that is true. But wouldn't we be adding after appending `cancel(fa/fd)`?

Comment (gxyd) [inline sympy/integrals/prde.py]: Okay, so should it be like `F` itself is given changed already as input (i.e `fa/fd` is already appending to `F` before input to `linear_relations_in_Q`) and then after that from the all the nullspace matrices we obtain the one put the coresponding `ri` of `fa/fd` to `1`?

Comment (jksuom) [inline sympy/integrals/prde.py]: The idea is to call this function for finding all rational linear relations and then choose those that satisfy some extra conditions such as one coefficient being  -1 as in the case you refer to, and possibly another being an integer. But the caller has to do that processing. `.nullspace` returns a (possibly empty) list of column matrices that are thought of as vectors. They form a basis of the nullspace. The final solutions are suitable linear combinations of the basis vectors.

(The relations can also be obtained from the rowspace, but then extra processing would be needed.)

**Comment by gxyd** [comment]
There are failures currently:
```python

[gxyd@stallman sympy{param-log-deriv}]$ python3 ./bin/test sympy/integrals/tests/test_rde.py 
============================================================= test process starts =============================================================
executable:         /home/gxyd/anaconda3/bin/python3  (3.6.0-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        67173900
hash randomization: on (PYTHONHASHSEED=4047578247)

sympy/integrals/tests/test_rde.py[10] ...Ef.....                                                                                         [FAIL]

_______________________________________________________________________________________________________________________________________________
____________________________________________ sympy/integrals/tests/test_rde.py:test_special_denom _____________________________________________
  File "/home/gxyd/foss/sympy/sympy/integrals/tests/test_rde.py", line 82, in test_special_denom
    Poly(1, t0), DE) == \
  File "/home/gxyd/foss/sympy/sympy/integrals/rde.py", line 248, in special_denom
    B = ba*pN.quo(bd) + Poly(n, DE.t)*a*derivation(p, DE).quo(p)*pN
  File "/home/gxyd/foss/sympy/sympy/core/decorators.py", line 91, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/foss/sympy/sympy/polys/polytools.py", line 4003, in __mul__
    return f.mul(g)
  File "/home/gxyd/foss/sympy/sympy/polys/polytools.py", line 1391, in mul
    _, per, F, G = f._unify(g)
  File "/home/gxyd/foss/sympy/sympy/polys/polytools.py", line 399, in _unify
    dom, lev = f.rep.dom.unify(g.rep.dom, gens), len(gens) - 1
  File "/home/gxyd/foss/sympy/sympy/polys/domains/domain.py", line 262, in unify
    return K0.unify_with_symbols(K1, symbols)
  File "/home/gxyd/foss/sympy/sympy/polys/domains/domain.py", line 240, in unify_with_symbols
    raise UnificationFailed("can't unify %s with %s, given %s generators" % (K0, K1, tuple(symbols)))
sympy.polys.polyerrors.UnificationFailed: can't unify ZZ with ZZ[x], given (t0, x) generators

================================= tests finished: 8 passed, 1 expected to fail, 1 exceptions, in 2.15 seconds =================================
DO *NOT* COMMIT!
```

**Comment by gxyd** [comment]
Fixed the above mentioned issue.

File: sympy/integrals/tests/test_prde.py [hunk 15]
```diff
@@ -52,7 +53,8 @@ def test_prde_special_denom():
     assert prde_special_denom(Poly(5*x*t + 1, t), Poly(t**2 + 2*x**3*t, t), Poly(t**3 + 2, t), G, DE) == \
         (Poly(5*x*t + 1, t), Poly(0, t), [(Poly(t, t), Poly(t**2, t)),
         (Poly(2*t, t), Poly(t, t))], Poly(1, x))
-    DE = DifferentialExtension(extension={'D': [Poly(1, x), Poly((t**2 + 1)*2*x, t)]})
+    DE = DifferentialExtension(extension={'D': [Poly(1, x), Poly((t**2 + 1)*2*x, t)]
```
Comment (gxyd) [inline sympy/integrals/tests/test_prde.py]: I am guessing we should have test cases for `parametric_log_deriv` corresponding to this (i.e whatever input `parametric_log_deriv` received for this test case)? Since this problem will eventually be tried (and in most of the cases be solved successfully) by "heuristic version" and never reach the "deterministic version" (after we put `try, except` block.

**Comment by czgdp1807** [comment]
@gxyd Are you still interested to work on it? In addition, I want to know which issue this PR was solving?

**Comment by czgdp1807** [comment]
Closing this for now. Please let us know if there is a need to re-open this. Thanks.

**Comment by oscarbenjamin** [comment]
@czgdp1807 is there an issue somewhere that tracks progress in Risch? We should reference these closed PRs there rather than just closing them. These represent substantial work that will be lost otherwise.
